### ISSUE TITLE
Execute rake tasks?

### ISSUE THREAD

**Issue by cmalpeli** [ondrejbartas/sidekiq-cron#133]
Is it possible for sidekiq-cron to execute rake tasks? Or should I convert them to Sidekiq Jobs?

**Comment by ondrejbartas** [comment]
Hi,

from my experience I am putting code into `Service` file as service pattern.
so in my rails app I have something like:

- app
  - models
  - services
    - some_super_cool_task.rb

which be executed from rake task, sidekiq worker or direclty in app (controller action etc).

Best thing from this aproach is that you can test that code.

Rake task is untestable

**Comment by marckohlbrugge** [comment]
I know this issue 4 years old, but I had the same question and ended up here. I figured I'd share my solution for others who can't easily convert their rake tasks to background jobs (for example rake tasks provided by other gems).

Create a generic background job or worker that can execute any rake task:

```ruby
# app/jobs/invoke_rake_task_job.rb

require "rake"
Rails.application.load_tasks

class InvokeRakeTaskJob < ApplicationJob
  def perform(args)
    Rake::Task[args["task"]].invoke(args["args"])
  end
end
``` 

In order to run the `sitemap:refresh` rake task add the following to your `config/schedule.yml`:

```yaml
sitemap_refresh:
  cron: 0 0 * * *
  class: InvokeRakeTaskJob
  args:
    task: sitemap:refresh
```

**Comment by romankovt** [comment]
> I know this issue 4 years old, but I had the same question and ended up here. I figured I'd share my solution for others who can't easily convert their rake tasks to background jobs (for example rake tasks provided by other gems).
> 
> Create a generic background job or worker that can execute any rake task:
> 
> ```ruby
> # app/jobs/invoke_rake_task_job.rb
> 
> require "rake"
> Rails.application.load_tasks
> 
> class InvokeRakeTaskJob < ApplicationJob
>   def perform(args)
>     Rake::Task[args["task"]].invoke(args["args"])
>   end
> end
> ```
> 
> In order to run the `sitemap:refresh` rake task add the following to your `config/schedule.yml`:
> 
> ```yaml
> sitemap_refresh:
>   cron: 0 0 * * *
>   class: InvokeRakeTaskJob
>   args:
>     task: sitemap:refresh
> ```

I want to add a note to this comment for anyone running rake tasks inside workers.

When you use `#invoke` method on a rake task, it will be executed **only once**. 

If you have `Rake::Task[args["task"]].invoke(args["args"])` as a sidekiq-cron job, you will have a task executed **for the first time** and it will just return nil on the next scheduled execution (unless you restarted sidekiq-cron process in between those scheduled executions)

You have 2 options:
1) Reenable task before executing it, something like:
```ruby
Rake::Task[args["task"]].reenable
Rake::Task[args["task"]].invoke(args["args"])
```
2) Use #execute instead of #invoke
```ruby
Rake::Task[args["task"]].execute(args["args"])
```
**Important!** #execute and #invoke are not the same! See more info: https://sampatbadhe.medium.com/rake-task-invoke-or-execute-419cd689c3bd
TL;DR if you have nested tasks or another task as a dependency, #invoke will execute the whole chain, whereas #execute executes only the specified task.

**Comment by marckohlbrugge** [comment]
@romankovt Thanks for sharing this Roman. I wasn't aware of this. I will update my code example in case someone misses your comment.

**Comment by camiloabrilth1999** [comment]
Hello! 
How can I put an environment variable in the args of config/sidekiq_scheduler.yml ?

**Comment by luilver** [comment]
Latest `Rails.application.load_tasks` already requires rake: 

```ruby
# Load Rake and railties tasks, and invoke the registered hooks.
# Check Rails::Railtie.rake_tasks for more info.
def load_tasks(app = self)
  require "rake"
  run_tasks_blocks(app)
  self
end
```

So we may avoid requiring twice ;-)

https://github.com/rails/rails/blob/main/railties/lib/rails/engine.rb#L470
